site/src/pages/WorkspacePage/WorkspaceActions/DownloadLogsDialog.tsx

```diff
@@ -0,0 +1,166 @@
+import { useTheme, type Interpolation, type Theme } from "@emotion/react";
+import Skeleton from "@mui/material/Skeleton";
+import { saveAs } from "file-saver";
+import JSZip from "jszip";
+import { useMemo, useState, type FC } from "react";
+import { useQueries, useQuery } from "react-query";
+import { agentLogs, buildLogs } from "api/queries/workspaces";
+import type { Workspace, WorkspaceAgent } from "api/typesGenerated";
+import {
+  ConfirmDialog,
+  type ConfirmDialogProps,
+} from "components/Dialogs/ConfirmDialog/ConfirmDialog";
+import { displayError } from "components/GlobalSnackbar/utils";
+import { Stack } from "components/Stack/Stack";
+
+type DownloadLogsDialogProps = Pick<
+  ConfirmDialogProps,
+  "onConfirm" | "onClose" | "open"
+> & {
+  workspace: Workspace;
+  download?: (zip: Blob, filename: string) => void;
+};
+
+type DownloadFile = {
+  name: string;
+  blob: Blob | undefined;
+};
+
+export const DownloadLogsDialog: FC<DownloadLogsDialogProps> = ({
+  workspace,
+  download = saveAs,
+  ...dialogProps
+}) => {
+  const theme = useTheme();
+  const agents = selectAgents(workspace);
+  const agentLogResults = useQueries({
+    queries: agents.map((a) => ({
+      ...agentLogs(workspace.id, a.id),
+      enabled: dialogProps.open,
+    })),
+  });
+  const buildLogsQuery = useQuery({
+    ...buildLogs(workspace),
+    enabled: dialogProps.open,
+  });
+  const downloadFiles: DownloadFile[] = useMemo(() => {
+    const files: DownloadFile[] = [
+      {
+        name: `${workspace.name}-build-logs.txt`,
+        blob: buildLogsQuery.data
+          ? new Blob([buildLogsQuery.data.map((l) => l.output).join("\n")], {
+              type: "text/plain",
+            })
+          : undefined,
+      },
+    ];
+
+    agents.forEach((a, i) => {
+      const name = `${a.name}-logs.txt`;
+      const logs = agentLogResults[i].data;
+      const txt = logs?.map((l) => l.output).join("\n");
+      let blob: Blob | undefined;
+      if (txt) {
+        blob = new Blob([txt], { type: "text/plain" });
+      }
+      files.push({ name, blob });
+    });
+
+    return files;
+  }, [agentLogResults, agents, buildLogsQuery.data, workspace.name]);
```

Comment:
Would we want to swap `agentLogResults` for `agentLogResults.data`? React Query objects have a lot of properties on them, so any one of them changing could re-trigger the memoized logic, even if the data itself is exactly the same